forem/forem
# Use a single listener for all follow buttons on a page

## What type of PR is this? (check all applicable)


## Description

We had a conversation on a previous profile preview card PR around the efficiency of individual click listeners for follow buttons - as the preview card work means that there are significantly more follow buttons on a given page, we're adding more and more individual click event listeners all to do the same thing. We agreed it would be more efficient to have a single click handler, which is what is implemented here.

 As part of the refactor I've moved the follow buttons code to JS packs so that we get the advantages that come with code splitting etc.

I've tried to be as thorough as possible in identifying all pages which need this new pack - to my knowledge the list is:

- Article page
- Podcast episode page
- Tag index page
- Individual tag page
- User profile page
- Organisation profile page
- Search results for users
- Notifications
- Sponsors (code existed to initialize follow buttons _if they are included in the sponsor HTML_, so I've made sure the pack is added here and tested locally by adding a sponsor with a follow button. I'm not sure how likely it is that any Forem other than DEV would know to add Follow buttons here though.)

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/222

## QA Instructions, Screenshots, Recordings

For each of the pages above, check that follow buttons:

- Display the correct initial state (e.g. 'Follow', 'Follow back', 'Following', 'Edit profile')
- Update correctly when you click them
- Trigger the `/follows` network request when you click them
- When you refresh the page, the updates are clearly still persisted

For the notifications page, you can trigger a follow notification by:

```rb
user = User.find_by(username: "admin_mcadmin")
other_user = User.first
follow = other_user.follows.create!(followable: user)
Notification.send_new_follower_notification_without_delay(follow)
```

For the sponsors page, you will need to set up the test by:

- Visiting Admin > Advanced > Sponsors
- Edit one of the pre-seeded sponsorships
- Set the status to "live" and check that the expiry date is in the future
- Add a follow button in the blurb HTML. You could use this one (just copied from an element in the DOM on another page):

```html
<button name="button" type="button" data-info="{&quot;id&quot;:2,&quot;className&quot;:&quot;User&quot;,&quot;style&quot;:&quot;&quot;}" class="crayons-btn follow-action-button whitespace-nowrap w-100 follow-user showing" data-verb="unfollow" >Follow</button>
```

- Save your changes and then when you visit the `/sponsors` page there should be a follow button to experiment with

**UPDATED**

I originally missed two edge cases that we should also check:

- When you are logged out, all follow buttons should result in the login modal being shown
- When there is more than one follow button for a user on the page (e.g. article author byline preview card, and sticky sidebar at the right), then clicking one should update both buttons

### UI accessibility concerns?

No direct impact, the buttons should all continue to be usable by both keyboard and mouse.

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
